File: src/controller/controller_veiculo.py
from model.veiculo import Veiculo
from conexions.conexion import Conexion


class Controller_Veiculo:
    def __init__(self):
        self.conn = Conexion('conexions/info/login.json', can_write=True)

    def inserir_veiculo(self) -> Veiculo:
        self.conn.connect()

        id_veiculo = input("ID do Veículo (Novo): ")

        if not self.verifica_existencia_veiculo(id_veiculo):
            tipo_veiculo = input("Tipo do Veículo (Novo): ")
            descricao_veiculo = input("Descrição do Veículo (Nova): ")

            self.conn.execute_non_query(f"""
                INSERT INTO Veiculo (ID_Veiculo, Tipo_Veiculo, Descricao_Veiculo) 
                VALUES (:1, :2, :3)
            """, (id_veiculo, tipo_veiculo, descricao_veiculo))

            df_veiculo = self.conn.sql_to_data_frame(f"SELECT * FROM Veiculo WHERE ID_Veiculo = {id_veiculo}")
            novo_veiculo = Veiculo(df_veiculo['id_veiculo'][0], df_veiculo['tipo_veiculo'][0], df_veiculo['descricao_veiculo'][0])
            print(novo_veiculo.to_string())
            return novo_veiculo
        else:
            print(f"O ID {id_veiculo} já está cadastrado.")
            return None

    def atualizar_veiculo(self) -> Veiculo:
        self.conn.connect()

        id_veiculo = input("ID do Veículo que deseja atualizar: ")

        if self.verifica_existencia_veiculo(id_veiculo):
            novo_tipo = input("Tipo do Veículo (Novo): ")
            nova_descricao = input("Descrição do Veículo (Nova): ")

            self.conn.execute_non_query(f"""
                UPDATE Veiculo 
                SET Tipo_Veiculo = :1, Descricao_Veiculo = :2 
                WHERE ID_Veiculo = :3
            """, (novo_tipo, nova_descricao, id_veiculo))

            df_veiculo = self.conn.sql_to_data_frame(f"SELECT * FROM Veiculo WHERE ID_Veiculo = {id_veiculo}")
            veiculo_atualizado = Veiculo(df_veiculo['id_veiculo'][0], df_veiculo['tipo_veiculo'][0], df_veiculo['descricao_veiculo'][0])
            print(veiculo_atualizado.to_string())
            return veiculo_atualizado
        else:
            print(f"O ID {id_veiculo} não existe.")
            return None

    def excluir_veiculo(self):
        self.conn.connect()

        id_veiculo = input("ID do Veículo que deseja excluir: ")

        if self.verifica_existencia_veiculo(id_veiculo):
            df_veiculo = self.conn.sql_to_data_frame(f"SELECT * FROM Veiculo WHERE ID_Veiculo = {id_veiculo}")
            self.conn.execute_non_query(f"DELETE FROM Veiculo WHERE ID_Veiculo = {id_veiculo}")
            veiculo_excluido = Veiculo(df_veiculo['id_veiculo'][0], df_veiculo['tipo_veiculo'][0], df_veiculo['descricao_veiculo'][0])
            print("Veículo Removido com Sucesso!")
            print(veiculo_excluido.to_string())
        else:
            print(f"O ID {id_veiculo} não existe.")

    def verifica_existencia_veiculo(self, id_veiculo: str) -> bool:
        df_veiculo = self.conn.sql_to_data_frame(f"SELECT * FROM Veiculo WHERE ID_Veiculo = {id_veiculo}")
        return not df_veiculo.empty

File: src/controller/controller_veiculo.rb
require_relative '../model/veiculo'
require_relative '../conexions/connection'
class Controller_veiculo

  def initialize
    config_file = 'conexions/info/login.json'
    @connection = Connection.new(config_file)
  end

  def inserir_veiculo
    database = @connection.connect
    collection = database.database[:veiculos]

    puts "Insira o ID do Veiculo:"
    STDOUT.flush
    id_veiculo = gets.chomp

    if verifica_existencia_veiculo(id_veiculo)
      puts "Insira o tipo de veiculo:"
      STDOUT.flush
      tipo_veiculo = gets.chomp
      puts "Insira a descrição de Veiculo:"
      STDOUT.flush
      descricao_veiculo = gets.chomp

      collection.insert_one({
                              "_id" => id_veiculo,
                              "tipo_veiculo" => tipo_veiculo,
                              "descricao_veiculo" => descricao_veiculo
                            })

      data = recupera_veiculo(id_veiculo)
      novo_veiculo = Veiculo.new(data[0]['_id'], data[0]['tipo_veiculo'], data[0]['descricao_veiculo'])
      puts novo_veiculo.to_s
      novo_veiculo
    else
      puts "O ID #{id_veiculo} já existe"
      nil
    end
  end

  def atualizar_veiculo
    database = @connection.connect
    collection = database.database[:veiculos]

    puts "Insira o ID do Veiculo:"
    STDOUT.flush
    id_veiculo = gets.chomp

    if verifica_existencia_veiculo(id_veiculo)
      puts "Insira o novo tipo do veiculo:"
      STDOUT.flush
      tipo_veiculo = gets.chomp
      puts "Insira a nova descrição do veiculo:"
      STDOUT.flush
      descricao_veiculo = gets.chomp

      collection.update_one({"_id" => id_veiculo}, {
        "$set" => {
          "tipo_veiculo" => tipo_veiculo,
          "descricao_veiculo" => descricao_veiculo
        }
      })

      data = recupera_veiculo(id_veiculo)
      novo_veiculo = Veiculo.new(data[0]['_id'], data[0]['tipo_veiculo'], data[0]['descricao_veiculo'])
      puts novo_veiculo.to_s
      novo_veiculo
    end
  end

  def excluir_veiculo
    database = @connection.connect
    collection = database.database[:veiculos]

    puts "Insira o ID do Veiculo que deseja excluir:"
    STDOUT.flush
    id_veiculo = gets.chomp
    if verifica_existencia_veiculo(id_veiculo)
      data = recupera_veiculo(id_veiculo)
      veiculo = Veiculo.new(data[0]['_id'], data[0]['tipo_veiculo'], data[0]['descricao_veiculo'])
      collection.delete_one({"_id" => id_veiculo})
      puts "Veículo excluído com sucesso"
      puts veiculo.to_s
    else
      puts "Veículo não existe, nenhuma alteração feita"
    end
  end

  def recupera_veiculo(id_veiculo)
    database = @connection.connect
    collection = database.database[:veiculos]
    collection.find({ "_id" => id_veiculo }).to_a
  end

  def verifica_existencia_veiculo(id_veiculo)
    result = recupera_veiculo(id_veiculo)
    !result.empty?
  end

end
